src/pgduckdb_xact.cpp

```diff
@@ -1,39 +1,172 @@
+#include "duckdb/common/exception.hpp"
 #include "pgduckdb/pgduckdb_duckdb.hpp"
 #include "pgduckdb/pgduckdb_utils.hpp"
+#include "pgduckdb/pg/transactions.hpp"
 
-extern "C" {
-#include "postgres.h"
-#include "access/xact.h" // RegisterXactCallback and XactEvent
+namespace pgduckdb {
+
+static int64_t duckdb_command_id = -1;
+static bool top_level_statement = true;
+
+/*
+ * Returns if we're currently in a transaction block. To determine if we are in
+ * a function or not, this uses the tracked top_level_statement variable.
+ */
+bool
+IsInTransactionBlock() {
+	return IsInTransactionBlock(top_level_statement);
```

Comment:
I assume this is safe to be called without a lock on `top_level_statement` because this will only be called in PG hooks, not in DDB threads?
(same for `duckdb_command_id` and potential `elog(ERROR` below?)

Reply:
I introduced a `pgduckdb::pg` and `pgduckdb::ddb` namespace now. I moved this to the `pgduckdb::pg` namespace. I think we should write down assumptions about using functions from the `pgduckdb::pg` namespace. The two main assumptions are imo:

1. They should be called from a Postgres hook, or while holding the global PG lock from a duckdb thread.
2. They do not throw elog errors, only exceptions

Reply:
Yeah I think that makes total sense.